extension/firefox/chrome/ShumwayCom.jsm

```diff
@@ -273,6 +291,98 @@ var ShumwayCom = {
           return;
         }
         onSystemResourceCallback = callback;
+      },
+      createLocalConnection: function(connectionName, callback) {
+        if (typeof connectionName !== 'string' ||
+            connectionName[0] !== '_' && connectionName.split(':').length !== 2) {
+          return -1; // LocalConnectionConnectResult.InvalidName
+        }
+        if (localConnectionsRegistry[connectionName]) {
+          return -2; // LocalConnectionConnectResult.AlreadyTaken
+        }
+        var connection = {
+          callback,
+          allowedSecureDomains: Object.create(null),
+          allowedInsecureDomains: Object.create(null)
+        };
+        localConnectionsRegistry[connectionName] = connection;
+        return 0; // LocalConnectionConnectResult.Success
+      },
+      hasLocalConnection: function(connectionName) {
+        return !!_getLocalConnection(connectionName);
+      },
+      closeLocalConnection: function(connectionName) {
+        if (!_getLocalConnection(connectionName)) {
+          return -1; // LocalConnectionCloseResult.NotConnected
+        }
+        delete localConnectionsRegistry[connectionName];
+        return 0; // LocalConnectionCloseResult.Success
+      },
+      sendLocalConnectionMessage: function(connectionName, methodName, argsBuffer, sender,
+                                           senderURL) {
+        if (typeof methodName !== 'string' || typeof senderURL !== 'string') {
+          return null;
+        }
+        var connection = _getLocalConnection(connectionName);
+        if (!connection) {
+          return null;
+        }
+        try {
+          var parsedURL = NetUtil.newURI(senderURL);
+          var senderDomain = parsedURL.host;
+          var allowedDomains = parsedURL.schemeIs('https') ?
```

Comment:
ooh, I see what you mean. So insecureDomains should always be a perfect subset of secureDomains? That'd be easy enough to do at the point where the domains are added by just adding the secure domains to both sets. It's not possible to revoke access, so we don't have to worry about removal from the sets.